### PR TITLE
update adroll to v2

### PR DESCRIPTION
**Pull request by sperand-io**
**Waiting on: https://github.com/segmentio/mongo-scripts/pull/10**

Port of https://github.com/segmentio/analytics.js-integrations/pull/616

> Three fixes rolled into one PR:
> 
> **1) Support for sending email (closes #610)**
> Original patch submitted by @leohklee. Thanks! :smile: 
> 
> **2) Fixes adroll custom data behavior (closes #588)**
> This was a great catch by @chiplay. As he explained in #588, the `__adroll.record_user` method parses out top level fields and "custom data" itself. The docs that show setting the `adroll_custom_data` object are referring to a global object on the window, not a sub-object that gets passed into that method. Not sure how we missed this :/
> 
> 
> 
> **3) Deprecates sending segments by name** ([Related Ticket](https://segment.zendesk.com/agent/tickets/28111))
> As of April it's required to use a [segmentId](https://help.adroll.com/hc/en-us/articles/203637220-Advanced-Non-URL-Segmentation-Strategies#content-anchor) instead of a segment name, so the strategy to deal with this in a non-breaking manner is to put new users on a "v2" and leave old users on "v1" (`options.legacy == true`).
> 
> Only distinction between v1 and v2 is that v1 would send unmapped events with `adroll_segments` set to the snake_cased event name, which will no longer be accepted for any new segments. The reason I'd argue we can't just remove support for unmapped events is because adroll still supports the `segment_name` approach for any segments created prior to March, and we have lots of existing customers relying on that who never needed to specifically map the events before unless they wanted to map multiple events to a given segment like the example here:
> 
>  
> 
> For _mapped events_, the behavior is the exact same, we'll just need to update documentation that new users need to map to the segment_id instead of the snaked name prepended with a `+`.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -97,17 +110,27 @@ AdRoll.prototype.track = function(track) {
   del(customProps, 'orderId');
   del(customProps, 'id');
   del(customProps, 'sku');
-  if (!is.empty(customProps)) data.adroll_custom_data = customProps;
+  if (!is.empty(customProps)) {
```
Comment (ndhoule) [inline lib/index.js]: Why do we need to check for emptiness? `each` won't do anything on an empty set.

File: lib/index.js [hunk 2]
```diff
@@ -97,17 +110,27 @@ AdRoll.prototype.track = function(track) {
...
+    each(customProps, function(key, val) {
```
Comment (ndhoule) [inline lib/index.js]: Let's use `foldl` here and produce a new object; I'd like to move toward using `each` only to produce side effects so that they'll stand out.

File: test/index.test.js [hunk 3]
```diff
@@ -102,166 +105,252 @@ describe('AdRoll', function() {
...
+
+        it('should track unnamed/categorized page', function() {
+          analytics.page({ url: 'http://localhost:34448/test/' });
+          analytics.called(window.__adroll.record_user, {
+            adroll_segments: 'loaded_a_page',
+            path: '/test/',
+            referrer: '',
+            title: 'integrations tests',
+            search: '',
+            url: 'http://localhost:34448/test/'
```
Comment (ndhoule) [inline test/index.test.js]: I realize that you're following precedent on these tests, but can you (let's do it in a separate PR) change these tests to not use hardcoded URLs/ports/title/etc.? Hardcoding these means that if we ever switch to a different test runner, change Duo's port number, run these tests in a different environment, use Mocha's `grep` querystring to whittle down tests. etc. etc. they'll all fail.

Comment (ndhoule) [inline test/index.test.js]: You'll have to do a small amount of research to make this all IE-compatible, but the `component/querystring` module includes most of it. (We might even just use it for this.)

File: test/index.test.js [hunk 4]
```diff
@@ -102,166 +105,252 @@ describe('AdRoll', function() {
...
-        adroll_custom_data: {
-          path: '/test/',
-          referrer: '',
-          title: 'integrations tests',
-          search: '',
-          name: 'Name',
-          category: 'Category',
-          url: 'http://localhost:34448/test/'
-        }
+      describe('V1', function() {
```
Comment (ndhoule) [inline test/index.test.js]: Can you split these into separate test files? (We can do that now.)

Comment (ndhoule) [inline test/index.test.js]: Any file in the `test/` directory named `*.test.js` will get picked up by the test runner.

File: lib/index.js [hunk 5]
```diff
@@ -20,9 +20,11 @@ var AdRoll = module.exports = integration('AdRoll')
   .global('__adroll_loaded')
   .global('adroll_adv_id')
   .global('adroll_custom_data')
+  .global('adroll_email')
   .global('adroll_pix_id')
   .option('advId', '')
   .option('pixId', '')
+  .option('v1', false)
```
Comment (ndhoule) [inline lib/index.js]: Would rather see a `version` option here.

File: lib/index.js [hunk 6]
```diff
@@ -97,17 +110,27 @@ AdRoll.prototype.track = function(track) {
...
-  if (!events.length) {
+  // no events found. only send this event if v1,
+  // as adroll no longer accepts segment by name
+  // and instead only by segment id
+  //
+  // new customers must map *all* events that they want
+  // sent to adroll in the interface
+  //
+  // if we are sending, snake case, since they only
+  // allowed segment names which were snake cased.
```
Comment (ndhoule) [inline lib/index.js]: wrap lines at 80 chars

File: test/index.test.js (hunk 4 above)
Comment (ndhoule) [inline test/index.test.js]: What's the difference between these two suites? Are all of the tests different? Just the event names?

File: lib/index.js (hunk 5 above)
Comment (sperand-io) [inline lib/index.js]: hmm... the key is that this is literally only to maintain backward compat with their old means of data collection (when event names were legal), whereas now they have IDs that we must map. 

so we _can't_ send them event names for new users, but _must_ for old users, which means there's no sense in actually exposing an option. it will be hidden and serve more as a "flag" than any sort of opt in change in functionality contract like the salesforce version bump.

that's why i originally called it `legacy`. 

totally open to switching to an option but thought you'd want the context first

File: lib/index.js (hunk 1 above)
Comment (sperand-io) [inline lib/index.js]: haha no idea, i wasn't thinking about refactor existing code at all to keep things _relatively_ sane in this change :)

File: lib/index.js (hunk 2 above)
Comment (sperand-io) [inline lib/index.js]: sounds good

File: lib/index.js (hunk 1 above)
Comment (ndhoule) [inline lib/index.js]: ah yeah, the difference being that we're iterating now where we weren't previously

File: lib/index.js (hunk 5 above)
Comment (ndhoule) [inline lib/index.js]: It's basically the same thing as this option—it's a flag that should not be exposed to the user. Basically suggestion is, rather than a `bool` it should be an `int`. Otherwise we're creating a singleton option; if we ever need to make another one of these changes, we have to create an entirely new option and migrate everyone instead of just grabbing `version=3`

Maybe we should get in the habit of prefixing internal options with an underscore, ala `_version`, to make it super clear.

File: test/index.test.js (hunk 3 above)
Comment (sperand-io) [inline test/index.test.js]: Sounds good. I wasn't looking at this update as an opportunity to rewrite the existing integration or its tests, so i just used the style that was there before. 

This update really was originally meant to be an urgent patch to stop sending them useless segment names for new customers (and since i figured it would be quick, I rolled in the other 2 fixes as well). I see now that was probably misguided on multiple fronts (conflating multiple issues into a single PR, ignoring flagrant issues with the existing base, etc :)

I can rewrite the tests, but I do feel strongly that doing so should be out of the immediate scope  and not hinder forward movement on this particular pull; I'm not adding any additional debt to what was already there, and this update has been an absurdly long time coming (which is my own fault) and needs to go out soon!

File: lib/index.js (hunk 5 above)
Comment (ndhoule) [inline lib/index.js]: Okay yeah, I'm calling it—let's start prefixing internal settings with a leading underscore.

File: test/index.test.js (hunk 3 above)
Comment (ndhoule) [inline test/index.test.js]: Yep, let's do it in a separate PR.

**Comment by ndhoule** [comment]
@sperand-io any update here?

**Comment by ndhoule** [comment]
@sperand-io light bump :car:

**Comment by sperand-io** [comment]
@ndhoule done. lmk what you think about the new test setup. now i've only broken out the tests where expected behavior is actually different, whereas [before](https://github.com/segment-integrations/analytics.js-integration-adroll/blob/8507a026ddd84d7d026e9e6baeaf1e9ea68ee61f/test/index.test.js) there were a lot of duplicate tests for v1 and v2 where expected behavior was the same. i'll either remove latest commit or rebase, just lmk which you think is better. thanks!

**Comment by hankim813** [comment]
@ndhoule @sperand-io can you guys ping me when this deploys?

File: lib/index.js [hunk 7]
```diff
@@ -78,36 +90,29 @@ AdRoll.prototype.page = function(page) {
...
-  del(customProps, 'orderId');
-  del(customProps, 'id');
-  del(customProps, 'sku');
-  if (!is.empty(customProps)) data.adroll_custom_data = customProps;
 
   each(events, function(event) {
-    // the adroll interface only allows for
-    // segment names which are snake cased.
-    data.adroll_segments = snake(event);
+    data.adroll_segments = event;
```
Comment (ndhoule) [inline lib/index.js]: The removal of `snake` here is going to mean events that were once sent in snake case may not be in the future, right? Seems like a v1 integration is still going to send these events, which will mess up existing customers' data?

**Comment by ndhoule** [comment]
> @ndhoule done. lmk what you think about the new test setup. now i've only broken out the tests where expected behavior is actually different, whereas before there were a lot of duplicate tests for v1 and v2 where expected behavior was the same. i'll either remove latest commit or rebase, just lmk which you think is better. thanks!

I would rather see these separated out into multiple files (e.g. `test/index.test.js`, `test/v1.test.js`, `test/v2.test.js`). I found the comingled suite difficult to read through/understand where to put a given test, which I worry will make maintaining tests difficult and error prone.

File: test/index.test.js [hunk 8]
```diff
@@ -238,30 +252,48 @@ describe('AdRoll', function() {
...
+          });
+
+          it('should snake_case and send unmapped events', function() {
+            analytics.track('Event A');
+            analytics.called(window.__adroll.record_user, {
+              adroll_segments: 'event_a'
+            });
+          });
+
+          it('should map revenue and order id', function() {
```
Comment (ndhoule) [inline test/index.test.js]: Seems like this test is duplicated above in `should pass in revenue and order id`. Break these tests apart into smaller tests; "should map revenue and order id" can likely be two individual tests. At that point your mapped test suite will just be a matter of testing whether or not the event will be renamed, which I'd expect will increase the amount of shared tests we have and make this a lot easier to maintain, which will quickly become a problem given the logic branch we've introduced by adding versioning.

File: lib/index.js [hunk 9]
```diff
@@ -78,36 +90,29 @@ AdRoll.prototype.page = function(page) {
...
     window.__adroll.record_user(data);
   });
 
-  // no events found
-  if (!events.length) {
+  // no events found. only send this event if v1, as adroll no longer
+  // accepts segments by name and instead only by segment id
+  if (this.options._version === 1 && !events.length) {
```
Comment (ndhoule) [inline lib/index.js]: Let's reformat this into an early return and move it above the `each` call above, and clarify the docs a bit—the relation between segment name -> segment ID and the events map was not clear to me when I first read this comment

``` js
if (!events.length) {
  // As of _____ Adroll no longer accepts segments by name, instead tracking by segment ID, which we represent as mapped events or whatever
  // Support legacy customers by sending unmapped events
  if (this.options._version === 1) {
    // make an adroll call
  }
  return;
}
```
